From b8e1f9fd02658c41bdc10b40c5fd670137276c42 Mon Sep 17 00:00:00 2001
From: Michael John <mj@tierone.software>
Date: Mon, 3 Aug 2026 09:47:40 -0400
Subject: [PATCH 28/29] xargs: don't call err(3) in the vfork'd child

Found by scan-build ('dangerous construct in a vforked process'):
the child called err(1, ...) on open(/dev/tty) or dup2 failure
between vfork() and execvp; err() invokes exit(), running atexit
handlers and stdio cleanup in the shared pre-exec address space.
Use warn(3) + _exit(1), preserving message and exit status.
---
 src.freebsd/findutils/xargs/xargs.c | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/src.freebsd/findutils/xargs/xargs.c b/src.freebsd/findutils/xargs/xargs.c
index 18b91fea..b218f225 100644
--- a/src.freebsd/findutils/xargs/xargs.c
+++ b/src.freebsd/findutils/xargs/xargs.c
@@ -607,8 +607,11 @@ exec:
 		break;
 	case 0:
 		if (oflag) {
-			if ((fd = open(_PATH_TTY, O_RDONLY)) == -1)
-				err(1, "can't open /dev/tty");
+			if ((fd = open(_PATH_TTY, O_RDONLY)) == -1) {
+				/* vfork'd child: warn(3) + _exit, not err(3). */
+				warn("can't open /dev/tty");
+				_exit(1);
+			}
 		} else if (aflag) {
 			/* don't redirect anything by default for -a */
 			fd = -1;
@@ -616,8 +619,11 @@ exec:
 			fd = open(_PATH_DEVNULL, O_RDONLY);
 		}
 		if (fd > STDIN_FILENO) {
-			if (dup2(fd, STDIN_FILENO) != 0)
-				err(1, "can't dup2 to stdin");
+			if (dup2(fd, STDIN_FILENO) != 0) {
+				/* vfork'd child: warn(3) + _exit, not err(3). */
+				warn("can't dup2 to stdin");
+				_exit(1);
+			}
 			close(fd);
 		}
 		execvp(argv[0], argv);
-- 
2.55.0

